internal/gcs/container.go

```diff
@@ -225,20 +245,83 @@ func (c *Container) Terminate(ctx context.Context) (err error) {
 // Wait waits for the container to terminate (or Close to be called, or the
 // guest connection to terminate).
 func (c *Container) Wait() error {
+	<-c.waitBlock
+	return c.waitError
+}
+
+// waitBackground waits for the GuestConnection to notify of container exit and
+// unblocks other Wait calls
+//
+// This MUST be called via a goroutine to wait on a background thread, and can
+// only be called once
+func (c *Container) waitBackground(ctx context.Context) {
+	var err error
+	ctx, span := trace.StartSpan(ctx, "gcs::Container::waitBackground")
+	defer span.End()
+	defer func() { oc.SetSpanStatus(span, err) }()
+	span.AddAttributes(trace.StringAttribute(logfields.ContainerID, c.id))
+
+	// wait for container exit
 	select {
-	case <-c.notifyCh:
-		return nil
 	case <-c.closeCh:
-		return errors.New("container closed")
+		// container in the guest was closed
+	case <-c.waitBlock:
+		// `c.Close()` was called
 	}
+
+	c.closeOnce.Do(func() {
+		// `notifyCh` and `closeCh` are closed by `GuestConnection`, so do not close here
+		close(c.waitBlock)
+		c.notifications.Close()
+	})
+
+	err = c.waitError
+	log.G(ctx).Debug("container exited")
 }
 
-func (c *Container) waitBackground() {
-	ctx, span := trace.StartSpan(context.Background(), "gcs::Container::waitBackground")
+// Notifications returns a list of notifications (including shutdown) about the container
+//
+// The Notifications queue will be closed by Close or when the container in the
+// quest exits
+func (c *Container) Notifications() (*queue.MessageQueue, error) {
+	return c.notifications, nil
+}
+
+// publishNotifications publishes notifications from the sent forwarded by
+// the GuestConnection. Currently only OOM events are published.
```

Comment:
This first sentence is confusing

Reply:
changed to: publishNotifications publishes notifications from the container in the guest that are forwarded by the GuestConnection.